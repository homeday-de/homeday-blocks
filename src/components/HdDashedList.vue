<template>
  <dl class="dashed-list">
    <template v-for="(item, index) in items">
      <h4
        v-if="item.title"
        :key="index"
        class="dashed-list__title"
      >
        {{ item.title }}
      </h4>

      <div
        v-for="({ label, value, secondaryValue }, subIndex) in (item.items || [item])"
        :key="`${index}${subIndex}`"
        class="dashed-list__item"
        :class="{
          'dashed-list__item--expanded-label': expandedLabel && !secondaryValue,
        }"
      >
        <dt>
          <span v-html="label" />
        </dt>
        <dd>
          <span v-html="value" />
        </dd>
        <dd v-if="secondaryValue">
          <div class="dashed-list__dash" />
          <span v-html="secondaryValue" />
        </dd>
      </div>
    </template>
  </dl>
</template>

<script>
export default {
  name: 'HdDashedList',
  props: {
    items: {
      type: Array,
      default: () => [],
    },
    expandedLabel: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<style lang="scss" scoped>
@import 'homeday-blocks/src/styles/mixins.scss';

$leaders-border: 1px dashed getShade($neutral-gray, 50);

.dashed-list {
  background-color: inherit;

  &__title {
    margin-top: $sp-l;
    margin-bottom: $sp-s;
    color: $neutral-gray;
    font-weight: 700;
  }

  &__dash {
    flex: 1;
    display: inline-flex;
    border-bottom: $leaders-border;
    margin-bottom: .47em;
    margin-left: 0.1em;
    margin-right: 0.1em;
  }
}

.dashed-list__item {
  display: flex;
  flex-flow: wrap;
  margin-bottom: $sp-s;

  > dt,
  > dd {
    position: relative;
    display: flex;
    align-items: flex-end;
  }

  > dt {
    width: 100%;
    color: getShade($neutral-gray, 80);
  }

  > dd {
    background-color: inherit;
    color: $cello-blue;
  }

  > dd + dd {
    flex: 1;
    justify-content: flex-end;
  }
}

@media (min-width: $break-tablet) {
  .dashed-list__item {
    > dt,
    > dd {
      overflow: hidden;
    }

    > dt {
      position: relative;
      width: 40%;
      max-width: 60%;
      margin-right: 0.1em;

      > span::after {
        content: "";
        position: absolute;
        border-bottom: $leaders-border;
        width: 100%;
        margin-left: 0.1em;
        margin-top: 1em;
      }
    }
  }

  .dashed-list__item--expanded-label {
    > dt {
      flex: 1;
      width: auto;
      max-width: unset;
    }
  }
}
</style>
