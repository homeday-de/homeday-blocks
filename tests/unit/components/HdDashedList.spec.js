import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdDashedList from '@/components/HdDashedList.vue';
import ITEMS from '@/stories/mocks/DASHED_LIST_ITEMS';

const ITEM_SELECTOR = '.dashed-list__item';
const LIST_WITH_EXPANDED_LABELS_CLASS = 'dashed-list--expanded-label';

const wrapperBuilder = wrapperFactoryBuilder(HdDashedList, {
  props: {
    items: ITEMS,
  },
});

describe('HdDashedList', () => {
  it('renders component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders items correctly', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.findAll(ITEM_SELECTOR).length).toEqual(4);
  });

  it('renders expanded labels if if the prop `expandedLabel` is true', async () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.classes()).not.toContain(LIST_WITH_EXPANDED_LABELS_CLASS);

    wrapper.setProps({
      expandedLabel: true,
    });

    await wrapper.vm.$nextTick();

    expect(wrapper.classes()).toContain(LIST_WITH_EXPANDED_LABELS_CLASS);
  });
});
