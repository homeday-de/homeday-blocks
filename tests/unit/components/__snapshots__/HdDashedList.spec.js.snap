// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdDashedList renders component 1`] = `
<dl class="dashed-list">
  <!---->
  <div class="dashed-list__item">
    <dt><span>Kaltmiete (wohnen) (pro Monat)</span></dt>
    <dd><span>1 (value)</span></dd>
    <!---->
  </div>
  <!---->
  <div class="dashed-list__item">
    <dt><span>Kaltmiete (Kosten) (Wohnen) (pro Monat)</span></dt>
    <dd><span>2 (value)</span></dd>
    <dd>
      <div class="dashed-list__dash"></div> <span>2 (secondaryValue)</span>
    </dd>
  </div>
  <h4 class="dashed-list__title">
    Title
  </h4>
  <div class="dashed-list__item">
    <dt><span>Item 3</span></dt>
    <dd><span>3</span></dd>
    <dd>
      <div class="dashed-list__dash"></div> <span>3 (secondaryValue)</span>
    </dd>
  </div>
  <div class="dashed-list__item">
    <dt><span>Item 4</span></dt>
    <dd><span>4</span></dd>
    <dd>
      <div class="dashed-list__dash"></div> <span>4 (secondaryValue)</span>
    </dd>
  </div>
</dl>
`;
